### PR TITLE
Fix segfault in get_ao_compression_ratio (#907)

### PR DESCRIPTION
**Pull request by dkovalev1**
Problem:
Segfault happened in the queries with 'get_ao_compression_ratio' function
in the target list and the argument for this function that was taken from 
a table. The problem was reproduced in the release build. In the development 
build the same queries led to assertion.

Cause:
Function 'get_ao_compression_ratio' executes a query on 'pg_aoseg.pg_aoseg_' 
table to calculate the compression ratio. A function that accesses 
hash-distributed or randomly distributed tables must be executed on 
coordinator. But the planner created plans with the 'get_ao_compression_ratio'
function which was executed on segments. Inside 'get_ao_compression_ratio' 
there was an assert to check this case, but on the release build the assert 
was removed. So, the code of the function was executed on a segment and got 
unexpected results from 'pg_aoseg.pg_aoseg_', that led to a segfault.

Fix:
Assert is changed to a runtime check. Now, if the planner creates a plan with
the function which is executed on segments, the user will get a proper error 
message. It is also done for the similar functions 'gp_update_ao_master_stats' 
and 'get_ao_distribution'. Plus, unreachable code is removed from internal 
functions.

Note: it seems that a fully proper way would be to mark these functions as 
PROEXECLOCATION_MASTER instead of PROEXECLOCATION_ANY, but that would require 
changing of the system catalog, and it is acceptable only between major 
versions update. So this approach is not used.

Note2: for an empty column-oriented table 'get_ao_compression_ratio' returns -1,
but for an empty row-oriented table 'get_ao_compression_ratio' returns 1. 
This behavior is inconsistent, but it is not altered by this commit as not 
related to the original issue. It has a reflection in the added test case. 
Be aware.

Changes comparing to the original commit:

1. Adopted parameters to GetAppendOnlyEntryAuxOids() in aosegfiles.c
2. Added ignore for warning "ignoring query cancel request for synchronous
replication to ensure cluster consistency." in tests
3. Removed gp_update_ao_master_stats which is not in 7.2

(cherry picked from commit 3dea58de92a9b040fe5ce194048a97f55bbdb71b)

Note: do not squash the commit to preserve authorship.

### CODE REVIEW

**Comment by RekGRpth** [comment]
use -x option with cherry-pick

File: src/test/regress/input/appendonly.source [hunk 1]
```diff
@@ -880,3 +880,38 @@ insert into fix_ao_truncate_last_sequence select 1, 1 from generate_series(1, 5)
...
+-- end_ignore
+
+create table t1_ao_comp_rate (a int) with (appendoptimized=true, orientation=row) distributed by (a);
+create table t2_ao_comp_rate (a int) with (appendoptimized=true, orientation=column) distributed by (a);
+create table tables_list (tname text) distributed by (tname);
+insert into tables_list values ('t1_ao_comp_rate'), ('t2_ao_comp_rate');
+
+-- Queries below should fail with a proper error message.
+
+-- It's possible that extected error cancels query on the some segment too early for it, after it started transaction
```
Comment (RekGRpth) [inline src/test/regress/input/appendonly.source]: > extected

?

**Comment by RekGRpth** [comment]
rebase to 7.2.0

**Comment by whitehawk** [comment]
Please fix the PR description, or, if commits shouldn't be squashed, all the commits's descriptions should be properly formatted and fully documented.

**Comment by whitehawk** [comment]
@dkovalev1  Please use -x option with cherry-pick to preserve authorship, add to the commit description what was changed comparing with the original commit and place a note in the PR description that commits shouldn't be squashed.

**Review by RekGRpth — APPROVED** [review]

**Review by whitehawk — APPROVED** [review]

**Review by whitehawk — APPROVED** [review]

**Review by RekGRpth — APPROVED** [review]

**Comment by dkovalev1** [comment]
bender build

**Pull request merged**